engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/VariableInstanceEntity.java

```diff
@@ -278,26 +275,52 @@ public TypedValue getTypedValue(boolean deserializeValue) {
     return typedValueField.getTypedValue(deserializeValue, isTransient);
   }
 
-  public void setValue(TypedValue value) {
-    // clear value fields
-    clearValueFields();
+  public void setValue(TypedValue newValue) {
+    clearValueFieldsBasedOnType(newValue);
 
-    typedValueField.setValue(value);
+    typedValueField.setValue(newValue);
+  }
+
+  public void clearValueFieldsBasedOnType(TypedValue newValue) {
+    if (variableWillChangeType(newValue)) {
+      clearValueFieldsBasedOnType(true);
+    } else {
+      clearValueFieldsBasedOnType(false);
+    }
   }
 
-  public void clearValueFields() {
+  public void clearValueFieldsBasedOnType(boolean deleteVariable) {
```

Comment:
🔧 Keep the existing name as we decide inside to delete based on the type:
```suggestion
  public void clearValueFields(boolean deleteVariable) {
```